src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -507,11 +505,28 @@ private void WriteResponseHeadersUnsynchronized(int streamId, int statusCode, Ht
     {
         try
         {
+            // In the case of the headers, there is always a status header to be returned, so BeginEncodeHeaders will not return BufferTooSmall.
             _headersEnumerator.Initialize(headers);
             _outgoingFrame.PrepareHeaders(headerFrameFlags, streamId);
-            var buffer = _headerEncodingBuffer.AsSpan();
+            _headerEncodingBuffer.ResetWrittenCount();
+            var buffer = _headerEncodingBuffer.GetSpan(_maxFrameSize)[0.._maxFrameSize]; // GetSpan might return more data that can result in a less deterministic behavior on the way headers are split into frames.
             var done = HPackHeaderWriter.BeginEncodeHeaders(statusCode, _hpackEncoder, _headersEnumerator, buffer, out var payloadLength);
-            FinishWritingHeadersUnsynchronized(streamId, payloadLength, done);
+            Debug.Assert(done != HPackHeaderWriter.HeaderWriteResult.BufferTooSmall, "Oversized frames should not be returned, because this always writes the status.");
+            if (done == HPackHeaderWriter.HeaderWriteResult.Done)
+            {
+                // Fast path, only a single HEADER frame.
+                _outgoingFrame.PayloadLength = payloadLength;
```

Comment:
Yes, `ResetWrittenCount` is invoked both here and before writing the trailers. But I would say the `Advance` is not needed because we know all headers fit in the `buffer` (which has the size of `_maxFrameSize`) and we know exactly the written payload length. So then:

- not need to split the buffer into multiple frames, it must fit in a single frame
- no need to use `_headerEncodingBuffer.WrittenSpan` because because the `buffer` can be sliced directly, so no need for `Advance`.

 (Initially I had an `Advance` call and used written span for uniformity)

Reply:
I was thinking of `GetSpan` and `Advance` as paired APIs, but I'm not very familiar with the `ArrayBufferWriter`.  However, I guess `Advance` only matters if you're trying to concatenate two spans which, as you point out, we're not trying to do here.  Makes sense.